enhancements/machine-config/admin-defined-node-disruption-policy.md

```diff
@@ -0,0 +1,296 @@
+
+## Proposal
+
+Create NodeDisruptionPolicy and NodeDisruptionPolicyStatus in the machineconfigurations.operator.openshift.io CRD, allowing specify user-provided NodeDisruptionPolicy, and view current version cluster defaults, as follows:
+
+```console
+apiVersion: operator.openshift.io/v1
+kind: MachineConfiguration
+spec:
+  nodeDisruptionPolicy:
+    files:
+      - path: "/etc/my-file"
+        actions:
+          - type: Reload
+            reload: 
+              serviceName: my.service
+          - type: DaemonReload
+status:
+  nodeDisruptionPolicyStatus:
+    files:
+      - path: "/etc/my-file"
+        actions:
+          - type: Reload
+            reload:
+              serviceName: my.service
+          - type: DaemonReload
+      - path: /etc/mco/internal-registry-pull-secret.json
+          actions:
+            - type: None
+        - path: /var/lib/kubelet/config.json
+          actions:
+            - type: None
+        - path: /etc/machine-config-daemon/no-reboot/containers-gpg.pub
+          actions:
+            - type: Reload
+              reload:
+                serviceName: crio.service
+        - path: /etc/containers/policy.json
+          actions:
+            - type: Reload
+              reload:
+                serviceName: crio.service
+        - path: /etc/containers/registries.conf
+          actions:
+            - type: Special
+    sshkeys:
+        - actions:
+          type: None
+```
+
+Which, when applied, will define actions for all future changes to these MachineConfig fields based on the diff of the update (rendered MachineConfigs).
+
+For this initial implementation we will support MachineConfig changes to:
+ - Files
+ - Units
+ - sshKeys
+And actions for:
+ - None
+ - Draining a node
+ - Reloading a service
+ - Restarting a service
+ - Daemon-reload and restarting of a service
+ - Reboot (default)
+ - Special (see workflow description)
+
+In the future, we can extend this in two ways:
+Supporting more changes/actions (e.g. user defined script to run after a certain change)
+Supporting diffing between ostree images (for image based updates) and bootc managed systems (for bifrost)
+
+
+### Workflow Description
+
+An admin of an OCP cluster can at any time create/update/delete nodeDisruptionPolicy fields in the MachineConfiguration object to reconfigure how they wish updates to be applied. There isn’t any immediate effect so long as the CR itself is valid.
+
+The MachineConfigOperator pod will process any update to the object and validate the NodeDisruptionPolicy for potential issues, although most, if not all are covered by API-level validation. The MCO pod will then merge the user provided contents to the cluster defaults (user contents will always override cluster defaults) and populate the nodeDisruptionPolicyStatus. Success and errors will also be reflected in the status:
```

Comment:
I suppose this could change in the future, but as it stands we are completely guarded by API validation, no?

Reply:
modifed based on our current implementation